nemo_curator/modules/dataset_ops.py

```diff
@@ -0,0 +1,150 @@
+import math
+from typing import Any, Callable, List, Optional
+
+import dask.dataframe as dd
+import numpy as np
+
+from nemo_curator.datasets.doc_dataset import DocumentDataset
+
+
+def default_filename(partition_num: int) -> str:
+    return f"file_{partition_num:010d}.jsonl"
+
+
+class Shuffle:
+    def __init__(
+        self,
+        seed: Optional[int] = None,
+        npartitions: Optional[int] = None,
+        partition_to_filename: Callable[[int], str] = default_filename,
+    ) -> None:
+        """
+        Randomly permutes the dataset. This will make the original "filename" column invalid, so if the column is present it will be overwritten.
+        Args:
+            seed: The random seed that will be used to determine which partition (file) each datapoint goes to.
+                Even with the seed set, the shuffle is not guaranteed to be deterministic if done in parallel.
+                Intiailize the Dask client with a single worker and single thread in order to ensure determinism.
+            npartitions: The output number of partitions to create in the dataset.
+                If None, it will retain the same number of partitions as the original dataset.
+            partition_to_filename: If the filename column is present, it will be overwritten.
+                Passing a function in through this argument allows the user to configure what the filename
+                will look like given the partition number. The default method names the partition
+                f'file_{partition_num:010d}.jsonl' and should be changed if the user is not using a .jsonl format.
+        """
+        self.seed = seed
+        self.npartitions = npartitions
+        self.partition_to_filename = partition_to_filename
+        self.rand_col = "_shuffle_rand"
+
+    def __call__(self, dataset: DocumentDataset) -> DocumentDataset:
+        new_npartitions = (
+            dataset.df.npartitions if self.npartitions is None else self.npartitions
+        )
+
+        rand_df = dataset.df.map_partitions(self._add_rand_col, new_npartitions)
+
+        shuffled_df = rand_df.shuffle(
+            self.rand_col, npartitions=new_npartitions, ignore_index=True
```

Comment:
The shuffle as mentioned makes the result non deterministic. While slower, I wonder if `set_index` or `sort_values` on the `rand_col` was running into a lot of issues at scale? 
If it's slower but works, it might be a "slow" alternative is reproducibility is important for folks.

Reply:
I never tested `set_index`. I definitely was aware of it, but I can't quite remember why I didn't try it. I think I was worried about destroying the user's existing index if that was useful, but I doubt that many users would do something like that Will it always be deterministic in its ordering?

Reply:
`set_index`/ `sort_values` will also sort by the `rand_col` therefore ensuring a more or less consistent distribution for the same seed (not sure if it's a stable sort which might slightly impact results if the same rand col values are repeated). 
It is however a lot more expensive and slower than a hash based shuffle

Reply:
Makes sense. From my testing it definitely is slower, but not by much. Only 26% slower on a dataset of 116 GB. I added another path for it though.